[V1] [Bugfix] eagle bugfix and enable correct lm_head for multimodal (2)

RonaldBXu opened the pull request:
This is a fixed version of #18034 . My previous code overwrote all model attributes, but we should only be overwriting the defaults - I had accidentally also overwrote the `'architectures'` attribute that was set a couple lines earlier.

@ekagra-ranjan @DarkLight1337 would appreciate a review, thanks!

<!--- pyml disable-next-line no-emphasis-as-heading -->

RonaldBXu committed:
bugfix and enable correct lm_head for multimodal

Signed-off-by: Ronald Xu <ronaldxu@amazon.com>

RonaldBXu committed:
Merge branch 'vllm-project:main' into llama4-eagle-fixes

RonaldBXu committed:
address comments

Signed-off-by: Ronald Xu <ronaldxu@amazon.com>

RonaldBXu committed:
Merge branch 'main' into llama4-eagle-fixes

RonaldBXu committed:
linter

Signed-off-by: Ronald Xu <ronaldxu@amazon.com>

RonaldBXu committed:
linter

Signed-off-by: Ronald Xu <ronaldxu@amazon.com>

RonaldBXu committed:
Merge branch 'main' into llama4-eagle-fixes

RonaldBXu committed:
fix ci failure

Signed-off-by: Ronald Xu <ronaldxu@amazon.com>

RonaldBXu committed:
Merge branch 'vllm-project:main' into llama4-eagle-fixes

ekagra-ranjan reviewed (approved):
LGTM - can you share where was the architectures overwritten and how was this [fixed](https://github.com/vllm-project/vllm/pull/18034#issuecomment-2903806313)?

RonaldBXu commented:
Previously, I had automatically overwrote the attributes of self with self.model. However, vllm adds the "Eagle" or "Eagle3" prefix to the architectures attribute a couple lines above (https://github.com/vllm-project/vllm/blob/main/vllm/transformers_utils/configs/eagle.py#L51), so LlamaForCausalLM -> EagleLlamaForCausalLM. So, my previous commit was overwriting this with the original architecture, which caused vllm to assign a target model cls to the draft model. This made it so we were initializing 2 target models so there was duplicate layer error thrown. Now, I respect the prefix addition so the draft model cls is initialized correctly.

I think when I was working on this the code for adding the "Eagle" or "Eagle3" prefix did not exist yet, so I had just modified the config.json file of the Eagle model itself to have the "Eagle" prefix. For example, I modified the architectures field manually in (https://huggingface.co/yuhuili/EAGLE-LLaMA3.1-Instruct-8B/blob/main/config.json) to be EagleLlamaForCausalLM. I guess vllm decided to incorporate the Eagle prefix at runtime rather than have an Eagle model's config.json point to the Eagle cls.
